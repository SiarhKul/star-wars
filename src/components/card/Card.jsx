import React, { useCallback } from "react";
import { getAbbreviation } from "../../utils";

export const Card = React.memo(
	({ uniqueName, card, BodyComponent, onUniqueQueryParam }) => {
		const setUniqueQueryParam = useCallback(() => {
			onUniqueQueryParam(card);
		}, []);

		
		return (
			<div className="card-wrapper">
				<div className="card" onClick={setUniqueQueryParam}>
					<div className="avatar">
						<div className="avatar__abbr">{getAbbreviation(uniqueName)}</div>
					</div>
					<BodyComponent card={card} />
				</div>
			</div>
		);
    }
);
