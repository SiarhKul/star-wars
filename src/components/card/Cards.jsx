//todo использовать useCallback
//todo убрать мультиклик на меню
//todo создать 2 функции
//todo вернуть влево карточки

import React, { useCallback } from "react";
import { getAbbreviation } from "../../utils";
import { useDispatch } from "react-redux";
import { isVisiblePopup } from "../../redux/actionsCreators/actionsCrators";
import { history } from "../../redux/reducers";

export const Cards = ({ name, path, contenCards, BodyComponent }) => {
	const dispatch = useDispatch();

	const getAbbreviationMemo = useCallback(
		cardName => getAbbreviation(cardName),
		[]
	);

	const isPopupVisibleMemo = useCallback(() => {
		dispatch(isVisiblePopup(true));
	}, []);

	const setUniqueQueryParam = card => {
		history.push({
			pathname: path,
			state: card,
			search: `?selected=${card[name]}`,
		});
	};

	return (
		<div className="cards-container">
			{contenCards.map(card => {
				return (
					<div className="card-wrapper" key={card[name]}>
						<div
							className="card"
							onClick={() => {
								setUniqueQueryParam(card);
								isPopupVisibleMemo();
							}}
						>
							<div className="avatar">
								<div className="avatar__abbr">
									{getAbbreviationMemo(card[name])}
								</div>
							</div>
							<BodyComponent card={card} />
						</div>
					</div>
				);
			})}
		</div>
	);
};
