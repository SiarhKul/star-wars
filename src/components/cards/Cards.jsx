//todo использовать useCallback
//todo убрать мультиклик на меню
//todo создать 2 функции
//todo вернуть влево карточки

import React, { useCallback } from "react";

import { history } from "../../redux/reducers";
import { Card } from "../card/Card";

export const Cards = ({ name, path, contenCards, BodyComponent }) => {
	const setUniqueQueryParam = useCallback(card => {
		history.push(`${path}/${card[name]}`, card);
	}, []);

	return (
		<div className="cards-container">
			{contenCards.map(card => (
				<Card
					onUniqueQueryParam={setUniqueQueryParam}
					key={card[name]}
					uniqueName={card[name]}
					card={card}
					path={path}
					BodyComponent={BodyComponent}
				/>
			))}
		</div>
	);
};
