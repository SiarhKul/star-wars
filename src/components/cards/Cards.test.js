import { shallow } from "enzyme";
import React from "react";
import { Card } from "../card/Card";
import { Cards } from "./Cards";

jest.mock("history", () => {
	return {
		createBrowserHistory: jest.fn(() => ({
			push: jest.fn(),
		})),
	};
});

describe("Test Cards component", () => {
	const cards = () => shallow(<Cards {...props} />);
	const props = {
		name: "name",
		path: "/planets",
		contenCards: [{ name: "Luke Skywalker" }],
		BodyComponent: jest.fn(),
	};
	let component;

	beforeEach(() => {
		component = cards();
	});

	it("should render component ", () => {
		expect(component).toMatchSnapshot();
	});

	it("Does 'Test' have component Cards", () => {
		expect(component.find(Card)).toHaveLength(1);
	});
});
