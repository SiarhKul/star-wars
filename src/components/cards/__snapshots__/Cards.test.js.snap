// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Cards component should render component  1`] = `
<div
  className="cards-container"
>
  <Memo()
    BodyComponent={[MockFunction]}
    card={
      Object {
        "name": "Luke Skywalker",
      }
    }
    key="Luke Skywalker"
    onUniqueQueryParam={[Function]}
    path="/planets"
    uniqueName="Luke Skywalker"
  />
</div>
`;
